### PR TITLE
[keccak] Alternative implementations

### PR DESCRIPTION
**Pull request by Brechtpd**
Description: https://hackmd.io/NaTuIvmaQCybaOYgd-DG1Q?view

Contains two (~three) different approaches to do keccak:
- bit implementation: does keccak on the bit level
- packed implementation: does keccak by packing multiple bits together and uses lookups to do the transformations. The `multi` variant can use multiple rows/internal round to reduce the number of columns, but because of that it is a little bit less efficient than the non `multi` variant (~5%) . This variant is similar to the current approach followed, but the implementation is quite a bit different.

The current code should be good enough to compare performance between the different implementations.

### CODE REVIEW

File: zkevm-circuits/src/keccak_circuit/keccak_bit.rs
Comment (CPerezz) [inline zkevm-circuits/src/keccak_circuit/keccak_bit.rs:220]: Any particular reason why this is still here?

File: zkevm-circuits/src/keccak_circuit/keccak_bit.rs
Comment (CPerezz) [inline zkevm-circuits/src/keccak_circuit/keccak_bit.rs:235]: ditto

File: zkevm-circuits/src/keccak_circuit/keccak_bit.rs
Comment (CPerezz) [inline zkevm-circuits/src/keccak_circuit/keccak_bit.rs:281]: Why not just use a `Fixed` or `Constant` for `q_end`?

File: zkevm-circuits/src/keccak_circuit/keccak_bit.rs
Comment (CPerezz) [inline zkevm-circuits/src/keccak_circuit/keccak_bit.rs:480]: As said above. Why can't this be also `fixed`?

File: zkevm-circuits/src/keccak_circuit/keccak_bit.rs
Comment (CPerezz) [inline zkevm-circuits/src/keccak_circuit/keccak_bit.rs:492]: With Han's query_challenge, this won't even be a parameter that is passed. But we can keep it for now.

**Review by CPerezz — COMMENTED** [review]
Review for the approach 1: **Bit solution**

So far it looks nice an simple. I would say a couple of things:
- The naming used is a bit poor and does not help to understand what's what (`b bits`, `c bits` etc...).
- The generation of the witness could be more automatic if we build methods as the `KeccakRow` generators for example as associated fns of `KeccakCircuit`.
- There are some commented code parts on `Theta` gate lookups generation which I'm not sure they should stay or are just missed things.


Also, as mentioned. We need to take into account that https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/659 will imply that the RLC is not longer a witness we provide to the circuit and rather just a value we "compute and can pass over to other circuits" via lookups.


Aside from all that. Awesome work with this @Brechtpd I'm super impressed! Will continue with the packed and multi_packed reviews. 
And from these, we should just pick the most performant and make some estimations to see if we should try to optimize more before going for the full impl with padding + table etc..

File: zkevm-circuits/src/keccak_circuit/keccak_bit.rs
Comment (Brechtpd) [reply zkevm-circuits/src/keccak_circuit/keccak_bit.rs:281]: I guess `q_end` is a bad name. The goal is for it to be used to mark the end of one keccak hash so a new keccak hash can be started. So where this will be set to 1 will change depending on the inputs.

File: zkevm-circuits/src/keccak_circuit/keccak_bit.rs
Comment (Brechtpd) [reply zkevm-circuits/src/keccak_circuit/keccak_bit.rs:492]: > We need to take into account that https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/659 will imply that the RLC is not longer a witness we provide to the circuit and rather just a value we "compute and can pass over to other circuits" via lookups.

I guess I'm not really sure how to best write these things until the challenge API is available. For things like the RLC encoded input bytes that can stretch a lot of rows, how should that be done so it is easily convertible to the new API? Even the calculations of the RLC encoded value of the keccak hash itself (which is the value exposed above) can be quite complex, so not sure if just writing the code now like there is some const randomness value makes sense.

Comment (han0110) [reply zkevm-circuits/src/keccak_circuit/keccak_bit.rs:492]: The `query_challenge` API will return `Value<F>` for usage, we can simply mock it with a `F::random()` for now (or some static random value). So I think currently the way this PR implemented should be easy to adapt.

**Comment by AronisAt79** [comment]
# zkevm-circuits PR#634 keccak bench results

## 1. packed_multi_keccak_bench

```
Cloud instance specs

Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   46 bits physical, 48 bits virtual
CPU(s):                          96
On-line CPU(s) list:             0-95
Thread(s) per core:              2
Core(s) per socket:              24
Socket(s):                       2
NUMA node(s):                    2
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           85
Model name:                      Intel(R) Xeon(R) Platinum 8275CL CPU @ 3.00GHz
Stepping:                        7
CPU MHz:                         2999.998
BogoMIPS:                        5999.99

RAM: 192 Gb
```

```
Result

test packed_multi_keccak::tests::bench_packed_multi_keccak_circuit_prover has been running for over 60 seconds
End:     Packed Keccak Multi Proof generation with 19 rows .........................390.956s
Start:   Keccak Proof verification
End:     Keccak Proof verification .................................................41.143ms
test packed_multi_keccak::tests::bench_packed_multi_keccak_circuit_prover ... ok

max RAM at ~ 35Gb
```

**Comment by AronisAt79** [comment]
@CPerezz @ChihChengLiang  bench results in previous comment ^^^^

**Comment by Brechtpd** [comment]
Awesome, thanks a lot for this @AronisAt79! Results are quite a bit differently than I expected.

Some extra context for these benches:

### 1. packed_multi_keccak_bench

Uses 5 rows/internal round  (assuming default settings were used, the main feature of this implementation is that the number of rows/internal rounds can be chosen) and so hashes 136 bytes/125 rows. In the bench `2**19/125 = 4194` keccak hashes are done in 391s giving **10.72 keccak hashes/s**.

### 2. packed_keccak_bench

Uses 1 row /internal round and so hashes 136 bytes/25 rows. In the bench `2**19/25 = 20971` keccak hashes are done in 2170s giving **9.66 keccak hashes/s**.  Surprises me that it's a bit slower than the packed_multi version...

### 3. bit_keccak_bench

Uses 1 row /internal round and so also hashes 136 bytes/25 rows. In the bench `2**19/25 = 20971` keccak hashes are done in 1819s giving **11.53 keccak hashes/s**.  This result is much closer to the packed implementations than I expected.


Pending and future prover optimizations should improve all of these quite a bit still.

**Comment by CPerezz** [comment]
@Brechtpd the only issue I see with the packed approaches is the big amount of RAM required. Considering specially that the bit approach performs better (at least now) than the others consuming less resources.

I'd also edit your prev message to say that they're not hashes/s but keccak_f's/s or permutations/s.
As the hash itself does not have a magnitude since the input can have any size.

Aside from that, as we commented in the past, we will check with the authors of Caulk whether it can help us to bring down these lookup costs and how could we integrate it within halo2.

File: zkevm-circuits/src/keccak_circuit/keccak_packed.rs
Comment (CPerezz) [inline zkevm-circuits/src/keccak_circuit/keccak_packed.rs:1111]: This will not even be needed in the future. You can (if you want) just use a `Expression::Constant(F::one())` and forget about passing the rlc and also comparing it to anything.

Also, the naming is not correct. This is the permutation hash. But definitely not the hash one. As we can have multiple calls to `keccak_f` per hash.

**Review by CPerezz — COMMENTED** [review]
Preliminary review over the `packed` approach for the keccak hashing.

Awesome job @Brechtpd !
I left some comments, lots of them repeated and dittos. 

There's only one thing that I don't fully understand which is that the naming of the structs and functions and also the comments seem to suggest that we're performing a hash. When indeed, we're just performing one `keccak_f` round of multiple ones that we can have per hash.

So as said in the previous PR with the naming, we should be a bit more precise. That also aplies to the `thin` and `fat` variable names repeated over and over through the code.

Once addressed, will review the final part (the tables).

File: zkevm-circuits/src/keccak_circuit/keccak_packed.rs
Comment (Brechtpd) [reply zkevm-circuits/src/keccak_circuit/keccak_packed.rs:183]: Each of these tables are made as large as the circuit height allows to pack as many bits as possible into a single lookup. So `normalize_3` will very likely contain many more bits than the `normalize_6` table, but of course with smaller bit values. So there is some overlap between the tables, but merging will not decrease the number of lookups I think.

File: zkevm-circuits/src/keccak_circuit/keccak_packed.rs
Comment (Brechtpd) [reply zkevm-circuits/src/keccak_circuit/keccak_packed.rs:186]: Same as above with the `normalize` tables, except for the `pack_unpack_table` (a bit less optimized for now so just works on bytes, but could be improved), each of these tables are made as big as possible to pack in as much as possible. A tag I think will not improve things because then the tables can pack less bits together, and each cell in the circuit only ever needs to do a single kind of lookup so no need to be able to switch between them.

File: zkevm-circuits/src/keccak_circuit/keccak_packed.rs
Comment (Brechtpd) [reply zkevm-circuits/src/keccak_circuit/keccak_packed.rs:327]: I know, something like the `expr`/`value` method of the utils would be much better.

File: zkevm-circuits/src/keccak_circuit/keccak_packed.rs
Comment (Brechtpd) [reply zkevm-circuits/src/keccak_circuit/keccak_packed.rs:504]: Yes exactly what you say. If the tables are able to do 8 bits/lookup, but we have e.g. two small parts of 4 bits we can use a single lookup instead. But yeah didn't do it because that gets a bit crazy.

File: zkevm-circuits/src/keccak_circuit/keccak_packed.rs
Comment (Brechtpd) [reply zkevm-circuits/src/keccak_circuit/keccak_packed.rs:546]: Oh this should have used `BIT_SIZE` instead, it's just the base.

File: zkevm-circuits/src/keccak_circuit/keccak_packed.rs
Comment (Brechtpd) [reply zkevm-circuits/src/keccak_circuit/keccak_packed.rs:582]: Less table space -> more lookups required -> more memory needed, I think is generally true. So I think nothing we can really do on the circuit to improve this further (except finding better ways to pack more stuff in fewer lookups of course).

Depending on where the peak memory is it may be easy to improve on the prover side (currently this is not handled very efficiently).

File: zkevm-circuits/src/keccak_circuit/keccak_packed.rs
Comment (Brechtpd) [reply zkevm-circuits/src/keccak_circuit/keccak_packed.rs:615]: I don't think many/large rotations have a big impact on the prover performance. The extended domain is currently x4 bigger, with https://github.com/privacy-scaling-explorations/halo2/pull/76 this will go down to just x2. This will give a 15% performance increase because FFTs will be smaller. So memory should also be improved quite a bit, but I haven't measured yet by how much.

I think it may also be possible to use copy constraints here instead, but I don't believe that will matter much.

File: zkevm-circuits/src/keccak_circuit/keccak_packed.rs
Comment (Brechtpd) [reply zkevm-circuits/src/keccak_circuit/keccak_packed.rs:664]: Okay this made sense to me but I don't really know where I picked up this naming from. :)

So "fat bits" are bits that can have bit values > 1. And "thin" bits are bits that are in the "normalized" range of [0,1].

File: zkevm-circuits/src/keccak_circuit/keccak_packed.rs
Comment (Brechtpd) [reply zkevm-circuits/src/keccak_circuit/keccak_packed.rs:842]: For sure, just there because it's very helpful while developing/optimizing to keep track of how many cells/lookups are required for the different steps.

File: zkevm-circuits/src/keccak_circuit/keccak_packed.rs
Comment (Brechtpd) [reply zkevm-circuits/src/keccak_circuit/keccak_packed.rs:846]: Yes, if the circuit height is > 2**16 can just do 3 column lookup table so 2 bytes can be processed/range checked at once.

File: zkevm-circuits/src/keccak_circuit/keccak_packed.rs
Comment (Brechtpd) [reply zkevm-circuits/src/keccak_circuit/keccak_packed.rs:871]: Sometimes cleaner to have a constraint builder per gate? But that doesn't always work well it kind of mixes both ways. To be honest though it's because it's currently not very well thought out to make the most sense.

File: zkevm-circuits/src/keccak_circuit/keccak_packed.rs
Comment (Brechtpd) [reply zkevm-circuits/src/keccak_circuit/keccak_packed.rs:927]: Maybe so. I created a helper function because this is used in a couple of places and was easier to change it (I actually had this wrong for some time).

File: zkevm-circuits/src/keccak_circuit/keccak_packed.rs
Comment (Brechtpd) [reply zkevm-circuits/src/keccak_circuit/keccak_packed.rs:1111]: I guess so, but it's only really used when in the last block for a certain hash so the name kind of makes sense no? Using the same name here as it would be called in the external interface for the keccak lookups would make sense (though don't know for sure what it is/will be).

**Comment by Brechtpd** [comment]
> @Brechtpd the only issue I see with the packed approaches is the big amount of RAM required. Considering specially that the bit approach performs better (at least now) than the others consuming less resources.

I believe the bit one should be the best performing one by quite a margin but for some reason it's only a little bit faster. Will need to look into that to understand why that is. However, the bit approach does use the most memory currently though!

> I'd also edit your prev message to say that they're not hashes/s but keccak_f's/s or permutations/s. As the hash itself does not have a magnitude since the input can have any size.

Done, as you can see I like to play it pretty loose with the naming :)

> Aside from that, as we commented in the past, we will check with the authors of Caulk whether it can help us to bring down these lookup costs and how could we integrate it within halo2.

Oh that's great.

**Comment by Brechtpd** [comment]
The implementations are now complete, but I will spend a bit more time cleaning up the code some more (and incorporating the feedback above).

File: zkevm-circuits/src/keccak_circuit/keccak_packed.rs
Comment (CPerezz) [reply zkevm-circuits/src/keccak_circuit/keccak_packed.rs:682]: I like it! Thanks for the explanation!!

File: zkevm-circuits/src/keccak_circuit/keccak_packed.rs
Comment (CPerezz) [reply zkevm-circuits/src/keccak_circuit/keccak_packed.rs]: Would be nice then to add some sort of comments on the top of the impl or similar. So that is easier to follow :)

File: zkevm-circuits/src/keccak_circuit/keccak_packed.rs
Comment (CPerezz) [reply zkevm-circuits/src/keccak_circuit/keccak_packed.rs]: I'm not sure either. But seemed a bit odd to me to have multiple CB's within the same context.

**Comment by CPerezz** [comment]
Hey @Brechtpd is it clean now so that we can go for a merge here? 
What do you think??

**Comment by Brechtpd** [comment]
Can be merged whenever you want for me. I will still do some changes to make things better, but like discussed that can be done in a follow up PR.

File: zkevm-circuits/src/keccak_circuit/util.rs
Comment (ChihChengLiang) [inline zkevm-circuits/src/keccak_circuit/util.rs:216]: maybe do debug assert here?

File: zkevm-circuits/src/keccak_circuit/util.rs
Comment (ChihChengLiang) [inline zkevm-circuits/src/keccak_circuit/util.rs:200]: Should this be EVM word(256 bits) or Keccak word(64 bits)?

File: zkevm-circuits/src/keccak_circuit/util.rs
Comment (ChihChengLiang) [inline zkevm-circuits/src/keccak_circuit/util.rs:214]: Does this "bit" have range 0 to 7? Can we name all of them something like "sparse bit" to distinguish from the binary bit?

**Review by ChihChengLiang — COMMENTED** [review]
I added some comments. @CPerezz Feel free to merge if we want to avoid conflict for halo2 upgrade. I still want to do more reviews, but I can add comments post-merge.

File: zkevm-circuits/src/keccak_circuit/util.rs
Comment (Brechtpd) [reply zkevm-circuits/src/keccak_circuit/util.rs:237]: Bit can be anywhere in [0,7] here, and will be converted to [0,1]. I added some comments.

File: zkevm-circuits/src/keccak_circuit/util.rs
Comment (Brechtpd) [reply zkevm-circuits/src/keccak_circuit/util.rs:258]: Most elements are added to the initial vec with `let mut part_sizes = vec![part_size; num_full_chunks];`.

File: zkevm-circuits/src/keccak_circuit/util.rs
Comment (Brechtpd) [reply zkevm-circuits/src/keccak_circuit/util.rs:204]: I think everywhere I talk about words they are keccak words (because keccak doesn't directly have anything to do with EVM words I think it should be clear from context, or at least I hope so).

File: zkevm-circuits/src/keccak_circuit/util.rs
Comment (Brechtpd) [reply zkevm-circuits/src/keccak_circuit/util.rs:218]: Bits here also anywhere in the [0,7] range, also added some comments here.

I think pretty much everywhere in the packed implementations "bit" means these sparse bits in the [0,7] range so not sure if calling it something else would be that useful because almost everywhere the new name would be used. Even in the bit based implementation the bits are >1 very often.

File: zkevm-circuits/src/keccak_circuit/util.rs
Comment (ChihChengLiang) [reply zkevm-circuits/src/keccak_circuit/util.rs:204]: The type `Word` here is `pub type Word = U256;` from `zkevm-circuits/eth_types`. Can we define a `type Word = u64` and replace U256 with it?

File: zkevm-circuits/src/keccak_circuit/util.rs
Comment (ChihChengLiang) [reply zkevm-circuits/src/keccak_circuit/util.rs:258]: Oh, makes sense now.

**Review by CPerezz — APPROVED** [review]
LGTM with the current status. 

I would merge this to prevent massive conflicts once the halo2 update PR is merged here.
I encourage anyway to keep the work of cleanup. And probably we should consider to start with the entire circuit as I attempted in #542

File: zkevm-circuits/src/keccak_circuit/util.rs
Comment (Brechtpd) [reply zkevm-circuits/src/keccak_circuit/util.rs:204]: Something >u64 is needed here because this will pack all 64 sparse bits into a single field element. With each sparse bit taking up 3 bits, the full packed word needs to be stored in at least `64*3 = 192` bits. Though I'm not sure anymore why I don't simply do all the math on a `Field` value directly, will see if it's easy to switch to that instead.

**Pull request merged**
